### ISSUE TITLE
sim:nxffs and sim:spiffs Test failure

### ISSUE THREAD

**Issue by btashton** [apache/incubator-nuttx#2092]
For the `sim:nxffs` configuration the self test is failing

For spiffs the logs are here [spiffs.txt](https://github.com/apache/incubator-nuttx/files/5433814/spiffs.txt)

For nxffs here is the first portion of the logs:
```
=== FILLING 1 =============================
Filled file system
  Number of files: 61
  Number deleted:  0
Directory:
 1. Type[8]: File  Name: JzjHDOyw5F6R/j8kmxeOM1/78Jzp
 2. Type[8]: File  Name: FdzG0jEy2NQlSI8RA/sg0qX9IDFExGR0x
 3. Type[8]: File  Name: SVjnrSFqb9CxFARVDgIxAUEvPUDaE4Ro4
 4. Type[8]: File  Name: BlhzQ9Br/d8RLulhpbOiFD6AbyPs/MBPt
 5. Type[8]: File  Name: DjDUaAna/ZJiqxuL7fqOJZLxb/NFyO/KI
 6. Type[8]: File  Name: SKVKqGgNrIcv9XdWMkfxI04QVNqzKPCC2
 7. Type[8]: File  Name: C
 8. Type[8]: File  Name: z7znO4HBnXIlDbjCay5qryGaVYffIsa9Q
 9. Type[8]: File  Name: 4WE00wK6I8CZWLKM0tu3mthuDiqooPR1z
10. Type[8]: File  Name: ZJlzbyZdrw0X7wukqiRW3vIDBpP3foLxA
11. Type[8]: File  Name: KAierUkgMiRd0omFrpHtXtoawi9Xo32Ej
12. Type[8]: File  Name: EnmzEwZ62DqOZOxMgE6N4ysXaVW0rnWni
13. Type[8]: File  Name: FReqW/q7PXIZXqUqibaP8YhiAxZDyTcab
14. Type[8]: File  Name: vGCorVt9z7MLSP5ufguJUUnZesvom0Vk1
15. Type[8]: File  Name: eGOW7bthGZsfwF
16. Type[8]: File  Name: AIsl8kQIy3j//jjqbpqRK8FCQJls8LoaD
17. Type[8]: File  Name: 0lPDaAqD8ndI2DP7l6SLTyhMKFE3Osa6S
18. Type[8]: File  Name: sch9TW6KdVK2sc6t9ZhlRuCI1cS2nZqR4
19. Type[8]: File  Name: RoB8bCDpdLOLYTFIJ8eqUtqlnzL4735Co
20. Type[8]: File  Name: uea
21. Type[8]: File  Name: YwtkcA8LqB0XHb6toMtWGmdyRPtg41CJe
22. Type[8]: File  Name: 5XF16qxUSMm5F3mhxIKPKbKzd4RuHj2P
23. Type[8]: File  Name: UyLbppM2F3ASuuaVVE/EzsSy8GQUsuLKi
24. Type[8]: File  Name: 6taxcnPNDbtOSl3qOUC1pGYThSDhrnvss
25. Type[8]: File  Name: hFjdeyXHle56PZAQ36D/iwo6OeV8QPO44
26. Type[8]: File  Name: 5DfR5f4Din2jehP/
27. Type[8]: File  Name: 2AXNn4WLjx1eYwvJ1RFRiOfsa7CP2tgxZ
28. Type[8]: File  Name: OB8hWs4qtRy5dWAAcAjnj2xu2Oe/F1EYR
29. Type[8]: File  Name: uLX9cO5a5PEi2Ra0X7WWy8fKiWNIgyW/V
30. Type[8]: File  Name: Yfub4Evmec4xPXspp/9CHYWc0TKFpEWJ
31. Type[8]: File  Name: 9HKRHjwFF0fmocZeSvzSLsVLSwp9aS0d6
32. Type[8]: File  Name: q9HE5YeXVimH1iqH9lDYGPe18SZCk0ASJ
ERROR: Failed to open file for reading: 28
  File name: /mnt/nxffs/EVwni/b4A/BqjDc42RCtKvjJ6DInifZ0tME26XGRXWl5ZAE
  File size: 2680
ERROR: Failed to read a file: 0
  File name: /mnt/nxffs/EVwni/b4A/BqjDc42RCtKvjJ6DInifZ0tME26XGRXWl5ZAE
  File size: 2680
ERROR: Failed to verify files
  Number of files: 61
  Number deleted:  0

=== DELETING 1 ============================
ERROR: Unlink 1 failed: 28
  File name:  /mnt/nxffs/7yee3fRJj3gXpKOEuQZMygaVcS3pSUUzIArhnEH7Bfv3yhg3G5HvbRezu06Xonr9JK
  File size:  1038
  File index: 7
ERROR: Unlink 1 failed: 28
  File name:  /mnt/nxffs/kPmXquTSrz
  File size:  7811
  File index: 8
ERROR: Unlink 1 failed: 28
  File name:  /mnt/nxffs/jpfZNjnGapJ
  File size:  2835
  File index: 9
ERROR: Unlink 1 failed: 28
  File name:  /mnt/nxffs/3Ki0kkm/8JgEHA/YMkXqdDJPL5PXR1LKqSZcZfelqZVfb7MTE
  File size:  2280
  File index: 10
ERROR: Unlink 1 failed: 28
  File name:  /mnt/nxffs/NrxbRqA9HY1JnY5yijLr5XvrklQyiA1EQlNK2
  File size:  3706
  File index: 11
ERROR: Unlink 1 failed: 28
  File name:  /mnt/nxffs/rRPpZVdXxJZFxc
  File size:  5998
  File index: 12
ERROR: Unlink 1 failed: 28
  File name:  /mnt/nxffs/Dr0dlhT/YF5uamAqw8AcNDvmDQIdOOvfnO549TdDfZdwNOYSR4b2o32IcYesXUnT8NStVcN1
  File size:  435
  File index: 13
ERROR: Unlink 1 failed: 28
  File name:  /mnt/nxffs/ycvIAKPGXU7EG6AdXTsCxrSiAmruJRy143K/j50nmvF9ENVIbqGis9O8jNUG
  File size:  5383
  File index: 14
ERROR: Unlink 1 failed: 28
  File name:  /mnt/nxffs/3WQeK17gTMTXBx3k/0RLXT7fFCStyffivV9u9P/PsDvwn0KccM0DqzOeTRyS5q0VrjC/ss5CC7yW9WROjjDMxogSov0c1StiiOIq5Blsq0W
  File size:  4242
  File index: 15
ERROR: Unlink 1 failed: 28
  File name:  /mnt/nxffs/0NzfziosNPM3OoHX6t3ED5ZY3iXKkTefv6eL3tBA64YX45A
  File size:  3349
  File index: 16
ERROR: Unlink 1 failed: 28
  File name:  /mnt/nxffs/jbcdNTqSQJs2s3hywoZyTtgDJmceVeI4Mfezqf4CKqoZKR/9wBjBA7nx7YsRqDYNzX/0BztrDpXQ4EybF9s0r36gitpy8YSHiORJfj
  File size:  7789
  File index: 17
ERROR: Unlink 1 failed: 28
  File name:  /mnt/nxffs/GZ0LvjwIHYnP3JtbMZ0OXyjgzwzAminV9BKRFewF/IlI1hLTW7n/hu7H9N02mVWR4cVzjdPAgx/kJZIL7SYN3EZrIqc0J5NRsEkmOZi/qiodV7
  File size:  3607
  File index: 18
ERROR: Unlink 1 failed: 28
  File name:  /mnt/nxffs/HJE1CxiqBFnzeuay1dqBakh9/NOsOu5dCw30U7LwGCztV2UKhCtYi3Wf9nLD1ze3UuTBS2LPean5cWe89eZmJPSABz
  File size:  3323
```

**Comment by btashton** [comment]
This should not have been closed quite yet.

**Comment by btashton** [comment]
I think there are very bad things happening here.  I cleaned up some configuration settings so the initial file writing works, but I am seeing what looks like inodes being overwritten.

We can see `EVwni0b4` verified, but after the next round of files are written it seems to have been overwritten.

```
Directory:
 1. Type[8]: File  Name: EVwni0b4
 2. Type[8]: File  Name: iDrgzlIPo4OF
 3. Type[8]: File  Name: Jr7euPt7BEA5R
 4. Type[8]: File  Name: txgGQpEb
 5. Type[8]: File  Name: 5lCi
 6. Type[8]: File  Name: PVd33
 7. Type[8]: File  Name: 8An2DVQZMEHq75SV9uzh
 8. Type[8]: File  Name: g0x7Sj4a8Vw1D0g6qmq
 9. Type[8]: File  Name: olO4Qw5g8KLnuD2DR3jZ
10. Type[8]: File  Name: QctLaNi7rqQi
11. Type[8]: File  Name: GXPLOKsvH0VAw
12. Type[8]: File  Name: aqPFCVxkRabuT
13. Type[8]: File  Name: m
14. Type[8]: File  Name: SUR5KNn9hLHhMEq2Bav
15. Type[8]: File  Name: c3s50WlUBrMjfvm7I6
16. Type[8]: File  Name: 8a
17. Type[8]: File  Name: pFM1IwB6GmecRJATcrbT
18. Type[8]: File  Name: WyD7w
19. Type[8]: File  Name: fD29PA0Z793z1eBJ
20. Type[8]: File  Name: l0TBNGiFmhLeSa
21. Type[8]: File  Name: K9UG0W0pUiI
22. Type[8]: File  Name: mDP5chi
23. Type[8]: File  Name: ufAsEPLpg
24. Type[8]: File  Name: FF9ZJ
25. Type[8]: File  Name: SY0qvj4Do8jsL3upQHIM
26. Type[8]: File  Name: pMZ5Ren
  Verifed file /mnt/nxffs/EVwni0b4
Deleted file 1 OK
Deleted file 2 OK
Deleted file 3 OK
Deleted file 4 OK
Deleted file 5 OK
Deleted file 6 OK
  Verifed file /mnt/nxffs/iDrgzlIPo4OF
Deleted file 8 OK
Deleted file 9 OK
Deleted file 10 OK
Deleted file 11 OK
Deleted file 12 OK
Deleted file 13 OK
Deleted file 14 OK
Deleted file 15 OK
Deleted file 16 OK
Deleted file 17 OK
Deleted file 18 OK
Deleted file 19 OK
Deleted file 20 OK
Deleted file 21 OK
Deleted file 22 OK
Deleted file 23 OK
Deleted file 24 OK
Deleted file 25 OK
Deleted file 26 OK
Deleted file 27 OK
Deleted file 28 OK
Deleted file 29 OK
Deleted file 30 OK
Deleted file 31 OK
Deleted file 32 OK
Deleted file 33 OK
Deleted file 34 OK
Deleted file 35 OK
Deleted file 36 OK
Deleted file 37 OK
Deleted file 38 OK
Deleted file 39 OK
Deleted file 40 OK
Deleted file 41 OK
Deleted file 42 OK
Deleted file 43 OK
Deleted file 44 OK
Deleted file 45 OK
Deleted file 46 OK
Deleted file 47 OK
Deleted file 48 OK
Deleted file 49 OK
Deleted file 50 OK
Deleted file 51 OK
Deleted file 52 OK
Deleted file 53 OK
Deleted file 54 OK
Deleted file 55 OK
Deleted file 56 OK
Deleted file 57 OK
Deleted file 58 OK
Deleted file 59 OK
Deleted file 60 OK
Deleted file 61 OK
Deleted file 62 OK
Deleted file 63 OK
Deleted file 64 OK
Deleted file 65 OK
Deleted file 66 OK
Deleted file 67 OK
  Verifed file /mnt/nxffs/Jr7euPt7BEA5R
  Verifed file /mnt/nxffs/ufAsEPLpg
Deleted file 70 OK
Deleted file 71 OK
  Verifed file /mnt/nxffs/txgGQpEb
Deleted file 73 OK
  Verifed file /mnt/nxffs/5lCi
  Verifed file /mnt/nxffs/PVd33
Deleted file 76 OK
Deleted file 77 OK
Deleted file 78 OK
Deleted file 79 OK
  Verifed file /mnt/nxffs/8An2DVQZMEHq75SV9uzh
  Verifed file /mnt/nxffs/FF9ZJ
  Verifed file /mnt/nxffs/g0x7Sj4a8Vw1D0g6qmq
  Verifed file /mnt/nxffs/SY0qvj4Do8jsL3upQHIM
Deleted file 84 OK
  Verifed file /mnt/nxffs/olO4Qw5g8KLnuD2DR3jZ
  Verifed file /mnt/nxffs/QctLaNi7rqQi
  Verifed file /mnt/nxffs/GXPLOKsvH0VAw
  Verifed file /mnt/nxffs/aqPFCVxkRabuT
  Verifed file /mnt/nxffs/m
  Verifed file /mnt/nxffs/SUR5KNn9hLHhMEq2Bav
  Verifed file /mnt/nxffs/c3s50WlUBrMjfvm7I6
  Verifed file /mnt/nxffs/8a
  Verifed file /mnt/nxffs/pFM1IwB6GmecRJATcrbT
  Verifed file /mnt/nxffs/WyD7w
  Verifed file /mnt/nxffs/fD29PA0Z793z1eBJ
  Verifed file /mnt/nxffs/l0TBNGiFmhLeSa
  Verifed file /mnt/nxffs/pMZ5Ren
  Verifed file /mnt/nxffs/K9UG0W0pUiI
  Verifed file /mnt/nxffs/mDP5chi
Verified!  Number of files: 26
  Number deleted:  0
End of loop memory usage:
VARIABLE  BEFORE   AFTER
======== ======== ========
arena     4000000  4000000
ordblks         7        9
mxordblk  3efb9b0  3efb9b0
uordblks   104530   103970
fordblks  3efbad0  3efc690
=== FILLING 3 =============================
  Created file /mnt/nxffs/rv1d23qQowAyOOybigf
  Created file /mnt/nxffs/IGm9i8Eeuk3xI5DO9j
  Created file /mnt/nxffs/k1UxeovUm
  Created file /mnt/nxffs/2dayoJq
  Created file /mnt/nxffs/u
  Created file /mnt/nxffs/N0TfRWdi8eP
  Created file /mnt/nxffs/vaELE4Vu9ujaN3mnceZ
  Created file /mnt/nxffs/svTtWKR6u32hPEKWJ
  Created file /mnt/nxffs/IXR
  Created file /mnt/nxffs/hDAqEP71e
  Created file /mnt/nxffs/MUMRVI0fvkxFYWwW0
  Created file /mnt/nxffs/F
  Created file /mnt/nxffs/EYa
  Created file /mnt/nxffs/rFSHIEYHDG1qCwlm3ft0
  Created file /mnt/nxffs/7WB0NPulsgooMAZ
  Created file /mnt/nxffs/XT0JXlcbHfqw0ctQbW7X
  Created file /mnt/nxffs/DIv21rB838p
  Created file /mnt/nxffs/C9aEwUyMf
  Created file /mnt/nxffs/S
  Created file /mnt/nxffs/tJYf0p
  Created file /mnt/nxffs/0NkL9XmJI
  Created file /mnt/nxffs/NCyEt6CsqiU
  Created file /mnt/nxffs/gtcKesg5H5mN0ddQ0Nl7
  Created file /mnt/nxffs/KZNZI
  Created file /mnt/nxffs/fBJQtXCFpvwkWCL1qbG10
  Created file /mnt/nxffs/LqtoYZOYVUoAJfg
  Created file /mnt/nxffs/i
  Created file /mnt/nxffs/eaAhsx0to7pMAE
  Created file /mnt/nxffs/9HXq5XP9MME43O
  Created file /mnt/nxffs/ubRz8G1I
  Created file /mnt/nxffs/x0X8ZxSe1
  Created file /mnt/nxffs/Q2rXBByVCiP7qOy
  Created file /mnt/nxffs/gqxCahyLpCg
  Created file /mnt/nxffs/ICNIvjBWk
  Created file /mnt/nxffs/g
  Created file /mnt/nxffs/vB1A
  Created file /mnt/nxffs/vbROd4I3x
  Created file /mnt/nxffs/c2XKKTVn0gDwuc
  Created file /mnt/nxffs/YD9HYOHRI
  Created file /mnt/nxffs/z2vy
  Created file /mnt/nxffs/q3v6J
  Created file /mnt/nxffs/9QFeGiONb8YOx
  Created file /mnt/nxffs/mhuEB
  Created file /mnt/nxffs/tMSfsPfCjANpL5z0Ybg
  Created file /mnt/nxffs/66P46Fl
  Created file /mnt/nxffs/RQRZthIu
  Created file /mnt/nxffs/4egmeJXMXKM
  Created file /mnt/nxffs/xgVKU5
  Created file /mnt/nxffs/PGGrd5Mp2VjX0Vg9Z
  Created file /mnt/nxffs/nuyctksR2i3NYCRW
  Created file /mnt/nxffs/AToBcySZDyj481
  Created file /mnt/nxffs/CnPmw
  Created file /mnt/nxffs/Gd2M
  Created file /mnt/nxffs/kfGMhZI9y5QvSmEvA4m0a
  Created file /mnt/nxffs/EMBHCBBN0sn41
  Created file /mnt/nxffs/6z642daOwdc6
  Created file /mnt/nxffs/EiCJh
  Created file /mnt/nxffs/kaiOFKV4XgC1F0sMF
  Created file /mnt/nxffs/xBKe5r7FNI
  Created file /mnt/nxffs/FA0frl9XCfVer
  Created file /mnt/nxffs/e2c547
  Created file /mnt/nxffs/LOX
  Created file /mnt/nxffs/orkByhdTUvnGLoO
  Created file /mnt/nxffs/A9eABe39d52Rj
  Created file /mnt/nxffs/41BD3vzzNn1AiVfA
  Created file /mnt/nxffs/DhZ5bn
  Created file /mnt/nxffs/JWlCZ5WnW2yN0FJqo
  Created file /mnt/nxffs/j1PojXROB
  Created file /mnt/nxffs/MK1bz
  Created file /mnt/nxffs/1C1V2pv
  Created file /mnt/nxffs/Tf0flG00CHZHf
  Created file /mnt/nxffs/Mq0tjUw0SI0
  Created file /mnt/nxffs/9y2ohs1rSvNbT55k3b0I
  Created file /mnt/nxffs/N62fy
Filled file system
  Number of files: 100
  Number deleted:  0
Directory:
 1. Type[8]: File  Name: tJYf0p
 2. Type[8]: File  Name: 0NkL9XmJI
 3. Type[8]: File  Name: NCyEt6CsqiU
 4. Type[8]: File  Name: gtcKesg5H5mN0ddQ0Nl7
 5. Type[8]: File  Name: KZNZI
 6. Type[8]: File  Name: fBJQtXCFpvwkWCL1qbG10
 7. Type[8]: File  Name: LqtoYZOYVUoAJfg
 8. Type[8]: File  Name: i
 9. Type[8]: File  Name: eaAhsx0to7pMAE
10. Type[8]: File  Name: 9HXq5XP9MME43O
11. Type[8]: File  Name: ubRz8G1I
12. Type[8]: File  Name: x0X8ZxSe1
13. Type[8]: File  Name: Q2rXBByVCiP7qOy
14. Type[8]: File  Name: gqxCahyLpCg
15. Type[8]: File  Name: ICNIvjBWk
16. Type[8]: File  Name: g
17. Type[8]: File  Name: vB1A
18. Type[8]: File  Name: vbROd4I3x
19. Type[8]: File  Name: c2XKKTVn0gDwuc
20. Type[8]: File  Name: YD9HYOHRI
21. Type[8]: File  Name: z2vy
22. Type[8]: File  Name: q3v6J
23. Type[8]: File  Name: 9QFeGiONb8YOx
24. Type[8]: File  Name: mhuEB
25. Type[8]: File  Name: tMSfsPfCjANpL5z0Ybg
26. Type[8]: File  Name: 66P46Fl
27. Type[8]: File  Name: RQRZthIu
28. Type[8]: File  Name: 4egmeJXMXKM
29. Type[8]: File  Name: xgVKU5
30. Type[8]: File  Name: PGGrd5Mp2VjX0Vg9Z
31. Type[8]: File  Name: nuyctksR2i3NYCRW
32. Type[8]: File  Name: AToBcySZDyj481
33. Type[8]: File  Name: CnPmw
34. Type[8]: File  Name: Gd2M
35. Type[8]: File  Name: kfGMhZI9y5QvSmEvA4m0a
36. Type[8]: File  Name: EMBHCBBN0sn41
37. Type[8]: File  Name: 6z642daOwdc6
38. Type[8]: File  Name: EiCJh
39. Type[8]: File  Name: kaiOFKV4XgC1F0sMF
40. Type[8]: File  Name: xBKe5r7FNI
41. Type[8]: File  Name: FA0frl9XCfVer
42. Type[8]: File  Name: e2c547
43. Type[8]: File  Name: LOX
44. Type[8]: File  Name: orkByhdTUvnGLoO
45. Type[8]: File  Name: A9eABe39d52Rj
46. Type[8]: File  Name: 41BD3vzzNn1AiVfA
47. Type[8]: File  Name: DhZ5bn
48. Type[8]: File  Name: JWlCZ5WnW2yN0FJqo
49. Type[8]: File  Name: j1PojXROB
50. Type[8]: File  Name: MK1bz
51. Type[8]: File  Name: 1C1V2pv
52. Type[8]: File  Name: Tf0flG00CHZHf
53. Type[8]: File  Name: Mq0tjUw0SI0
54. Type[8]: File  Name: 9y2ohs1rSvNbT55k3b0I
55. Type[8]: File  Name: N62fy
ERROR: Failed to open file for reading: 2
  File name: /mnt/nxffs/EVwni0b4
  File size: 6394
ERROR: Failed to read a file: 0
  File name: /mnt/nxffs/EVwni0b4
  File size: 6394
ERROR: Failed to verify files
  Number of files: 100
  Number deleted:  0
```

**Comment by btashton** [comment]
This is also highly suspicious `EVwni0b4` should be the first inode.  It looks like this is getting totally corrupted when the new files are written.
Before the writing of the new files
```
NXFFS Dump:
  BLOCK:OFFS  TYPE  STATE   LENGTH
      0:0     BLOCK ERASED    512
      1:5     INODE OK       6394
      1:39    DATA  OK        463
      2:5     DATA  OK        497
      3:5     DATA  OK        497
```
After the writing of the new files
```
NXFFS Dump:
  BLOCK:OFFS  TYPE  STATE   LENGTH
      0:5     INODE OK       3283
      0:32    DATA  OK        470
      1:5     DATA  OK        497
      2:5     DATA  OK        497
      3:5     DATA  OK        497
      4:5     DATA  OK        497
      5:5     DATA  OK        497
      6:5     DATA  OK        117
      6:132   DATA  OK        211
```

**Comment by btashton** [comment]
When nxffs_mediacheck is called on block 0 and the filesystem looks like this, it will fail thinking that there are no inodes.  This is not correct.  It kind of looks like `nxffs_nextentry` is hitting that empty space and not looking far enough to see block 1 before giving up and saying the rest of the flash must be zero.

```
NXFFS Dump:
  BLOCK:OFFS  TYPE  STATE   LENGTH
      0:0     BLOCK ERASED    512
      1:5     INODE DELETED  6394
      1:39    DATA  OK        463
      2:5     DATA  OK        497
      3:5     DATA  OK        497
      4:5     DATA  OK        497
      5:5     DATA  OK        497
      6:5     DATA  OK        497
      7:5     DATA  OK        497
      8:5     DATA  OK        497
      9:5     DATA  OK        497
     10:5     DATA  OK        497
     11:5     DATA  OK        497
     12:5     DATA  OK        497
     13:5     DATA  OK        464
     13:479   INODE DELETED   339
```
If I force `nxffs_mediacheck` to start at block 1 instead of 0 this will pass.

**Comment by patacongo** [comment]
I am not sure if this related, but there is documented SPIFFS bug in the TODO list (should be a GIT issue):

```
  Title:       SILENT SPIFFS FILE TRUNCATION
  Description: Under certain corner case conditions, SPIFFS will truncate
               files.  All of the writes to the file will claim that the
               data has been written but after the file is closed, it may
               be a little shorter than expected.

               This is due to how the caching is implemented in SPIFFS:

               1. On each write, the data is not written to the FLASH but
                  rather to an internal cache in memory.
               2. When the a write causes the cache to become full, the
                  content of cache is flushed to memory.  If that flush
                  fails because the FLASH has become full, write will
                  return the file system full error (ENOSPC).
               3. The cache is also flushed when the file is closed (or
                  when fsync() is called).  These will also fail if the
                  file system becomes full.

               The problem is when the file is closed, the final file
               size could be smaller than the number of successful writes
               to the file.

               This error is probably not so significant in a real world
               file system usage:  It requires that you write continuously
               to SPIFFS, never deleting files or freeing FLASH resources
               in any way.  And it requires the unlikely circumstance that
               the final file written has its last few hundred bytes in
               cache when the file is closed but there are even fewer bytes
               available on the FLASH.  That would be rare with a cache
               size of a few hundred bytes and very large serial FLASH.

               This issue does cause the test at apps/testing/fstest to
               fail.  That test fails with a "Partial Read" because the
               file being read is smaller than number bytes written to the
               file.  That test does write small files continuously until
               file system is full and even the the error is rare.  The
               boards/sim/sim/sim/configs/spiffs test can used to
               demonstrate the error.
  Status:      Open
  Priority:    Medium.  It is certain a file system failure, but I think that
               the exposure in real world uses cases is very small.
```

I think that the real world exposure of this problem is rare since it depends on a condition where the physical media becomes full while there is still cached data.  However, the spiffs test does show this failure.  It would probably go away with a different random seed, but I wanted to keep the problem in place and repeatable so that it could be fixed sometime.

I think that the fix would be to reserve sectors that are at least the size of the cache so that this condition could not occur.

**Comment by patacongo** [comment]
This bug is in the SPIFFS base code.  It would be interesting to see if they solved it.  The SPIFFS project has gone a different way, however, now focusing on security.  So nothing would be easily backported.

**Comment by btashton** [comment]
I think the issue is different. This happens when pack is required, it does not find the first inode because it finds all the empty flags in the first block.  Shouldn't it be looking at the start of blocks not just incrementing byte by byte? 

This only happens once we have written and deleted enough that we need to circle back to the beginning to reclaim space marked deleted.

**Comment by btashton** [comment]
I will have to look later today, but this seems suspect https://github.com/apache/incubator-nuttx/pull/1431  That might explain why we have the empty block in the first place.

**Comment by btashton** [comment]
Looks like my analysis was correct here.  The linked PR did cause the first block to be skipped when formatting the filesystem which caused the pack routine to fail.  That PR needs to be reverted or fixed, this will cause corruption of the filesystem.

**Comment by patacongo** [comment]
> 
> 
> Looks like my analysis was correct here. The linked PR did cause the first block to be skipped when formatting the filesystem which caused the pack routine to fail. That PR needs to be reverted or fixed, this will cause corruption of the filesystem.

These sim tests are very valuable.  We should encourage a suite of sim tests to verify all major features.  That would be a first step toward some automated regression testing.

Regression testing is sorely needed.  NuttX is gaining a reputation for being unstable.  Alan pointer this out to me:  https://www.embarcados.com.br/blink-led-no-esp32-com-o-rtos-nuttx/#comment-10549  It is in Portuguese, but translates to:

"Professional readers who may be curious about using NuttX, don't bother.  The latest versions are fundamentally broken, the howtos are out of date.  The email list is practically dead… it's a shame because it was a very interesting concept .. a posix in a small footprint..."

**Comment by patacongo** [comment]
What would it take to automate sim:nxffs in CI?  Would it be (1) building the sim, (2) running the test, and (2) checking for a successful result in the output?

**Comment by btashton** [comment]
Yeah this is something that I mentioned in an email a while back about the 10.0.0 release. I wrote some scripting already using the Robot Framework that starts to run some of these tests and it's my intention to have it in place for the next release. We also have the LTP work that I intended to wire up.

This is actually one of the reasons that the release got delayed, I ran all the sim configurations as well as all the configurations for multiple boards and found multiple issues. It's my intention to run all these again this weekend as well as the 15 or so supported boards that I have. There is a spreadsheet I shared that I used.

Getting these to be part of that suite of tests is also something I want to get working as well.

Additionally I have a webapp that I started (it's in my GitHub) that takes the build artifacts from our CI runs and makes it really easy for people to test them and submit an issue against a given build. I'm hoping that this will make it easy for people to periodically run a quick test against a board they have.  It also will do some automatic work like keeping track of the build size over time.
